src/provider/p_ibmca.c

```diff
@@ -0,0 +1,1392 @@
+                     IBMCA_LOGDIR, prov_name, provctx->debug_pid)
+                                        >= (int)sizeof(debug_file)) {
+            put_error_ctx(provctx, IBMCA_ERR_INTERNAL_ERROR,
+                          "IBMCA_LOGDIR too long: '%s'", IBMCA_LOGDIR);
+            return 0;
+        }
+
+        provctx->debug_file = fopen(debug_file, "a");
+        if (provctx->debug_file == NULL) {
+            put_error_ctx(provctx, IBMCA_ERR_INTERNAL_ERROR,
+                          "Failed to open trace file: '%s': %s", debug_file,
+                          strerror(errno));
+            return 0;
+        }
+
+        fprintf(provctx->debug_file,
+                "*** Trace activated for IBMCA version %s ***\n",
+                IBMCA_VERSION);
+    }
+
+    return 1;
+}
+
+static int ibmca_config_fips(struct ibmca_prov_ctx *provctx,
+                             const char *key, const char *value)
+{
+    if (ibmca_config_bool(provctx, key, value, &provctx->fips) == 0)
+        return 0;
+
+    provctx->fips_configured = true;
+    return 1;
+}
+
+static int ibmca_config_algorithms(struct ibmca_prov_ctx *provctx,
+                                   const char *key, const char *value)
+{
+    char *val, *tok;
+    int i, rc = 1;
+    bool found;
+
+    ibmca_debug_ctx(provctx, "provctx: %p key: '%s' value: '%s'",
+                    provctx, key, value);
+
+    val = P_STRDUP(provctx, value);
+    if (val == NULL) {
+        put_error_ctx(provctx, IBMCA_ERR_MALLOC_FAILED,
+                      "Failed to strdup config value");
+        return 0;
+    }
+
+    for (i = 0; ica_mech_infos[i].algo != NULL; i++)
+        provctx->algo_enabled[i] = false;
+
+    for (tok = strtok(val, ","); tok != NULL; tok = strtok(NULL, ",")) {
+        ibmca_debug_ctx(provctx, "algorithm: '%s'", tok);
+
+        if (strcasecmp(tok, IBMCA_CONFIG_ALGO_ALL) == 0) {
+            for (i = 0; ica_mech_infos[i].algo != NULL; i++)
+                provctx->algo_enabled[i] = true;
+
+            continue;
+        }
+
+        for (i = 0, found = false; ica_mech_infos[i].algo != NULL; i++) {
+            if (strcasecmp(tok, ica_mech_infos[i].algo) == 0) {
+                provctx->algo_enabled[i] = true;
+                found = true;
+                break;
+            }
+        }
+        if (found)
+            continue;
+
+        put_error_ctx(provctx, IBMCA_ERR_CONFIGURATION,
+                      "Unknown algorithm name '%s' in configuration for provider '%s'",
+                      tok, provctx->name);
+        rc = 0;
+        break;
+    }
+
+    P_FREE(provctx, val);
+    return rc;
+}
+
+static int ibmca_get_configuration(struct ibmca_prov_ctx *provctx)
+{
+    OSSL_PARAM params[NUM_CONFIG_ITEMS + 1];
+    char *param_ptrs[NUM_CONFIG_ITEMS];
+    const char *value;
+    int i, num;
+
+    ibmca_debug_ctx(provctx, "provctx: %p", provctx);
+
+    for (num = 0; ica_mech_infos[num].algo != NULL; num++)
+        ;
```

Comment:
Yes, but I would like to keep this a NULL-terminated list anyway. It might happen in the future that ica_mech_infos becomes an external variable and then the size would not be to determine that easy outside of this module. For consistency, I would like to keep the size determination as is also here.